### PR TITLE
feat(cat-voices): wallet link service

### PR DESCRIPTION
**Pull request by dtscalac**
# Description
- Refactors wallet link to make sure cubits don't communicate with the wallet extension directly but rather through a service.
- Refactors error handling, introduces a RegistrationException thrown by RegistrationService and LocalizedRegistrationException that can be mapped to a user-displayable error message.

## Related Issue(s)

Refers #818

## Please confirm the following checks

* [x] My code follows the style guidelines of this project
* [x] I have performed a self-review of my code
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation
* [x] My changes generate no new warnings
* [x] I have added tests that prove my fix is effective or that my feature works
* [x] New and existing unit tests pass locally with my changes
* [x] Any dependent changes have been merged and published in downstream module

### CODE REVIEW

File: catalyst_voices/packages/catalyst_voices_models/lib/src/registration/exception/registration_exception.dart [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import 'package:equatable/equatable.dart';
+
+/// A base exception thrown during user registration.
+abstract base class RegistrationException
```
Comment (damian-molinski) [inline catalyst_voices/packages/catalyst_voices_models/lib/src/registration/exception/registration_exception.dart:4]: looks like `sealed` class

File: catalyst_voices/packages/catalyst_voices_services/lib/src/registration/registration_service.dart [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+/// Manages the user registration.
+final class RegistrationService {
+  final TransactionConfigRepository _configRepository;
+
+  const RegistrationService(this._configRepository);
+
+  /// Returns the available cardano wallet extensions.
+  Future<List<CardanoWallet>> getCardanoWallets() {
+    return CatalystCardano.instance.getWallets();
```
Comment (damian-molinski) [inline catalyst_voices/packages/catalyst_voices_services/lib/src/registration/registration_service.dart:15]: `CatalystCardano` should be injected, just like `TransactionConfigRepository` is

File: catalyst_voices/packages/catalyst_voices_services/lib/src/registration/registration_service.dart [hunk 3]
```diff
@@ -0,0 +1,107 @@
+import 'package:catalyst_cardano/catalyst_cardano.dart';
+import 'package:catalyst_cardano_serialization/catalyst_cardano_serialization.dart';
+import 'package:catalyst_voices_models/catalyst_voices_models.dart';
+import 'package:catalyst_voices_repositories/catalyst_voices_repositories.dart';
+import 'package:catalyst_voices_services/src/registration/registration_transaction_builder.dart';
+
+/// Manages the user registration.
+final class RegistrationService {
+  final TransactionConfigRepository _configRepository;
```
Comment (damian-molinski) [inline catalyst_voices/packages/catalyst_voices_services/lib/src/registration/registration_service.dart:9]: `_transactionConfigRepository `

File: catalyst_voices/packages/catalyst_voices_services/lib/src/registration/registration_service.dart [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+      return registrationBuilder.build();
+    } catch (error) {
+      if (error is RegistrationException) {
+        rethrow;
+      }
+
+      throw const RegistrationUnknownException();
+    }
+  }
```
Comment (damian-molinski) [inline catalyst_voices/packages/catalyst_voices_services/lib/src/registration/registration_service.dart:70]:
```dart
    try {
      final walletApi = await wallet.enable();

      final registrationBuilder = RegistrationTransactionBuilder(
        transactionConfig: await _configRepository.fetch(networkId),
        networkId: networkId,
        seedPhrase: seedPhrase,
        roles: roles,
        changeAddress: await walletApi.getChangeAddress(),
        rewardAddresses: await walletApi.getRewardAddresses(),
        utxos: await walletApi.getUtxos(
          amount: Balance(
            coin: CardanoWalletDetails.minAdaForRegistration,
          ),
        ),
      );

      return registrationBuilder.build();
    } 
    on RegistrationException {
      rethrow;
    } catch (error) {
      throw const RegistrationUnknownException();
    }

```

File: catalyst_voices/packages/catalyst_voices_view_models/lib/src/registration/exception/localized_registration_exception.dart [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+  /// Creates a subclass of [LocalizedException] matching the [exception].
+  factory LocalizedRegistrationException.from(
+    RegistrationException exception,
+  ) {
+    if (exception is RegistrationInsufficientBalanceException) {
+      return const LocalizedRegistrationInsufficientBalanceException();
+    } else if (exception is RegistrationTransactionException) {
+      return const LocalizedRegistrationTransactionException();
+    }
```
Comment (damian-molinski) [inline catalyst_voices/packages/catalyst_voices_view_models/lib/src/registration/exception/localized_registration_exception.dart:19]: with sealed class refactor to `switch`

File: catalyst_voices/packages/catalyst_voices_models/lib/src/registration/exception/registration_exception.dart (hunk 1 above)
Comment (dtscalac) [reply catalyst_voices/packages/catalyst_voices_models/lib/src/registration/exception/registration_exception.dart]: fixed

**Review by damian-molinski — APPROVED** [review]
LGTM

**Pull request merged**